Add save to Readwise Reader integration to Hacker News extension

russellyeo opened the pull request:
## Description

Added optional support for saving hacker news articles to [Readwise Reader](https://readwise.io/read).

### Features
- Securely add your API key via preferences
- A new Action "Save to Readwise Reader" becomes available after you have added your API key
- Performing the action will save the URL to your Readwise Reader account via the [Readwise Reader API](https://readwise.io/reader_api)
- Previously saved articles will show up in the list with a `SavedDocument` icon

### Screenshots

<img width="862" alt="Screenshot 2025-01-31 at 22 02 50" src="https://github.com/user-attachments/assets/17d98a66-3b18-489a-aa4b-6dc1ad0cf94a" />

<img width="862" alt="Screenshot 2025-01-31 at 22 04 32" src="https://github.com/user-attachments/assets/35920a40-6529-4914-81fa-9c53e0f26233" />

<img width="862" alt="Screenshot 2025-01-31 at 22 03 34" src="https://github.com/user-attachments/assets/19372696-4e01-45c3-9db3-3005929cd5a0" />

<img width="862" alt="Screenshot 2025-01-31 at 21 56 51" src="https://github.com/user-attachments/assets/cb4d8ceb-b480-4a2a-8165-b79766515d4b" />


## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

pernielsentikaer commented on extensions/hacker-news/src/preferences.ts:
```diff
@@ -0,0 +1,34 @@
+import { getPreferenceValues } from "@raycast/api";
+
+interface Preferences {
+  /** Readwise API token from https://readwise.io/access_token */
+  readwiseToken: string;
+}
```
You don't need to manually set preference types as this is autogenerated in `raycast-env.d.ts`x

russellyeo replied:
Ah cool! Thanks for the feedback

pernielsentikaer reviewed (changes requested):
Looks like a nice update, could you check the suggestion I made 😊

pernielsentikaer commented:
Could you look into the merge conflicts too?

Also, we're testing @greptileai which also wants to do a review, could you look into that it finds if it makes sense?

pernielsentikaer commented on extensions/hacker-news/src/preferences.ts:
```diff
@@ -0,0 +1,29 @@
+ * @returns The configured Readwise API token
+ * @throws Error if the token is not configured
+ */
+export function getReadwiseToken(): string {
+  const preferences = getPreferenceValues();
+  if (!preferences.readwiseToken) {
+    throw new Error("Readwise API token not configured");
+  }
+  return preferences.readwiseToken;
```
You can add <Preferences> since we have `raycast-env.d.ts` available 😊

pernielsentikaer commented:
@greptileai can you check again 🙂

pernielsentikaer commented:
Can you show me a screencast of with and without the Readwise token? I don't have a token to test with 🫠

russellyeo commented on extensions/hacker-news/src/frontpage.tsx:
```diff
@@ -74,6 +75,13 @@ function Actions(props: { item: Parser.Item }) {
+          <Action
+            icon={Icon.SaveDocument}
+            title="Save to Readwise Reader"
+            onAction={() => saveToReadwise(props.item.link || "")}
```
Super cool that the bot knows about the components in the Raycast libraries! 

The `saveToReadwise` function does do error handling internally and shows a success/error toast message, but I guess it would be better software design to separate the API action and the UI. I'll do that now.

russellyeo commented:
@pernielsentikaer Thanks for fixing the changelog and preferences! 

I reviewed the feedback from the bot and made a couple of improvements to the code:
- Moved the success/failure toast logic into the frontpage component for better separation of concerns (UI in one file, API request in another)
- Changed `getReadwiseToken()` to return an optional string (string | undefined) instead of throwing an error, and updated usages to handle the undefined case appropriately
- Simplified `hasReadwiseToken()` to use `getReadwiseToken()` function

I also made an improvement to ensure the saved icon appears immediately by loading the saved URLs into useState and refreshing it after successful saves.

Overall the bot is pretty cool, but it seems to have made up some things (there's no `launchCommand` function in the code?) and some suggestions weren't that useful. However, it did make me think about other code smells in the same area, so it was still useful feedback!

russellyeo commented:
Here's a screencast of the changes in action 🙂

https://github.com/user-attachments/assets/b164b25b-5bc9-45d5-8b61-36179596a9c7

russellyeo commented:
@greptileai can you review again please? 🙂

russellyeo commented on extensions/hacker-news/src/frontpage.tsx:
```diff
@@ -45,19 +76,31 @@ function setTitle(title: string, topic: Topic) {
+  onSave: (url: string) => Promise<void>;
+}) {
+  // Check if this item is in the savedUrls array
+  const isSaved = props.item.link ? (props.savedUrls.includes(props.item.link) || isUrlSaved(props.item.link)) : false;
```
Yeah fair point, probably good enough to check the local state only

russellyeo commented on extensions/hacker-news/src/readwise.ts:
```diff
@@ -0,0 +1,115 @@
+  if (!urls.includes(url)) {
+    urls.push(url);
+    cache.set(CACHE_KEY, JSON.stringify(urls));
+  }
```
Good idea, added

russellyeo commented on extensions/hacker-news/CHANGELOG.md:
```diff
@@ -1,5 +1,10 @@
 # Hacker News Changelog
 
+## [🚀 Add Readwise Reader integration] - {PR_MERGE_DATE}
```
`PR_MERGE_DATE` ❤️

russellyeo commented on extensions/hacker-news/src/readwise.ts:
```diff
@@ -0,0 +1,155 @@
+const cache = new Cache();
+const CACHE_KEY = "readwise-saved-urls";
+// Maximum number of URLs to store in the cache
+const MAX_CACHE_SIZE = 100;
```
Added a cache limit here. 100 is probably more than enough.

pernielsentikaer reviewed (approved):
Hi 👋 

Looks good to me, approved 🔥

In the next version, we can always configure it to delete the oldest one when it reaches the limit, I suppose

pernielsentikaer commented:
No 🙂
